### DISCUSSION TITLE
2 Problems : Timezone & AP mode

### DISCUSSION THREAD

**Discussion by belu40** [Hieromon/AutoConnect#454] [How can I?] [answered]
Hello!
i read a lot of closed and solved issues but i didn't understand well how to solve 2 main problems that i have with AutoConnect.
I state that i think this is a great library and permits to have an easier life using esp modules! But i still have some question for Hieromon and the comunity...
In my project (code below) i started from the Simple example from Hieromon with connection to NTP server and made some tests...

problem1: I need to save the timezone setting made on the timezone page to reload the same setting after shutting down my esp32 and  repower it. Now i found that the timezone setting is lost every time the esp is shutted down. (the dates and time that i see in my main page of the browser shows a date and time around 1970 after every restart. I made some tests to create a txt file saved with spiffs with the timezone configuration and try to read it at every restart but with no success( see code below)

problem2: I have problems with the access point mode. When i first power on ESP32 after programming it, it's correctly configured as access point and with a mobile device a set the correct information (settings) of my wireless network. Then when the network is not available anymore i need to get the access point mode working again after a set timeout but i can't get this working. I tried also to change the autoReconnect option from true to false but didn't understood well how this function works.

I hope to solve this! Thank you in advance!
below the code!

#if defined(ARDUINO_ARCH_ESP8266)
#include <ESP8266WiFi.h>
#include <ESP8266WebServer.h>
#elif defined(ARDUINO_ARCH_ESP32)
#include <WiFi.h>
#include <WebServer.h>
#endif
#include <time.h>
#include <AutoConnect.h>
#include <SPIFFS.h>

String tz;
String tz1;

static const char AUX_TIMEZONE[] PROGMEM = R"(
{
  "title": "Data e Ora",
  "uri": "/timezone",
  "menu": true,
  "element": [
    {
      "name": "caption",
      "type": "ACText",
      "value": "Impostazioni Data e Ora locali",
      "style": "font-family:Arial;font-weight:bold;text-align:center;margin-bottom:10px;color:DarkSlateBlue"
    },
    {
      "name": "timezone",
      "type": "ACSelect",
      "label": "Selezionare Zona",
      "option": [],
      "selected": 10
    },
    {
      "name": "newline",
      "type": "ACElement",
      "value": "<br>"
    },
    {
      "name": "start",
      "type": "ACSubmit",
      "value": "OK",
      "uri": "/start"
    }
  ]
}
)";

typedef struct {
  const char* zone;
  const char* ntpServer;
  int8_t      tzoff;
} Timezone_t;

static const Timezone_t TZ[] = {
  { "Europe/London", "europe.pool.ntp.org", 0 },
  { "Europe/Berlin", "europe.pool.ntp.org", 1 },
  { "Europe/Helsinki", "europe.pool.ntp.org", 2 },
  { "Europe/Moscow", "europe.pool.ntp.org", 3 },
  { "Asia/Dubai", "asia.pool.ntp.org", 4 },
  { "Asia/Karachi", "asia.pool.ntp.org", 5 },
  { "Asia/Dhaka", "asia.pool.ntp.org", 6 },
  { "Asia/Jakarta", "asia.pool.ntp.org", 7 },
  { "Asia/Manila", "asia.pool.ntp.org", 8 },
  { "Asia/Tokyo", "asia.pool.ntp.org", 9 },
  { "Australia/Brisbane", "oceania.pool.ntp.org", 10 },
  { "Pacific/Noumea", "oceania.pool.ntp.org", 11 },
  { "Pacific/Auckland", "oceania.pool.ntp.org", 12 },
  { "Atlantic/Azores", "europe.pool.ntp.org", -1 },
  { "America/Noronha", "south-america.pool.ntp.org", -2 },
  { "America/Araguaina", "south-america.pool.ntp.org", -3 },
  { "America/Blanc-Sablon", "north-america.pool.ntp.org", -4},
  { "America/New_York", "north-america.pool.ntp.org", -5 },
  { "America/Chicago", "north-america.pool.ntp.org", -6 },
  { "America/Denver", "north-america.pool.ntp.org", -7 },
  { "America/Los_Angeles", "north-america.pool.ntp.org", -8 },
  { "America/Anchorage", "north-america.pool.ntp.org", -9 },
  { "Pacific/Honolulu", "north-america.pool.ntp.org", -10 },
  { "Pacific/Samoa", "oceania.pool.ntp.org", -11 }
};

#if defined(ARDUINO_ARCH_ESP8266)
ESP8266WebServer Server;
#elif defined(ARDUINO_ARCH_ESP32)
WebServer Server;
#endif

AutoConnect       Portal(Server);
AutoConnectConfig Config;       // Enable autoReconnect supported on v0.9.4
AutoConnectAux    Timezone;
////////////////////////////////////////////////////////////////////////////////////////////ROOT PAGE_NTP_VISUALIZZA_DATI_MACCHINA
void rootPage() {
  String  content =
    "<html>"
    "<head>"
    "<meta name=\"viewport\" content=\"width=device-width, initial-scale=1\">"
    "<script type=\"text/javascript\">"
    "setTimeout(\"location.reload()\", 1000);"
    "</script>"
    "</head>"
    "<body>"
    "<h2 align=\"center\" style=\"color:black;margin:20px;\">H228 Industry 4.0</h2>"
    "<h3 align=\"center\" style=\"color:gray;margin:10px;\">{{DateTime}}</h3>"
    "<p style=\"text-align:center;\">Reload the page to update the time.</p>"
    "<p></p><p style=\"padding-top:15px;text-align:center\">" AUTOCONNECT_LINK(COG_24) "</p>"
    "</body>"
    "</html>";
  static const char *wd[7] = { "Dom","Lun","Mar","Mer","Gio","Ven","Sab" };
  struct tm *tm;
  time_t  t;
  char    dateTime[26];

  t = time(NULL);
  tm = localtime(&t);
  sprintf(dateTime, "%04d/%02d/%02d(%s) %02d:%02d:%02d.",
    tm->tm_year + 1900, tm->tm_mon + 1, tm->tm_mday,
    wd[tm->tm_wday],
    tm->tm_hour, tm->tm_min, tm->tm_sec);
  content.replace("{{DateTime}}", String(dateTime));
  Server.send(200, "text/html", content);
  Serial.println ("dateTime");
  Serial.println (dateTime);
  
}

void startPage() {
  // Retrieve the value of AutoConnectElement with arg function of WebServer class.
  // Values are accessible with the element name.
    tz = Server.arg("timezone");
 
  Serial.println("tz");
  Serial.println(tz);

  for (uint8_t n = 0; n < sizeof(TZ) / sizeof(Timezone_t); n++) {
      String tzName = String(TZ[n].zone);
    if (tz.equalsIgnoreCase(tzName)) {
      configTime(TZ[n].tzoff * 3600, 0, TZ[n].ntpServer);
      Serial.println("Time zone: " + tz);
      Serial.println("ntp server: " + String(TZ[n].ntpServer));
      break;     
    }
    Serial.println ("tzName");
    Serial.println (tzName);
    
  }

  // The /start page just constitutes timezone,
  // it redirects to the root page without the content response.
  Server.sendHeader("Location", String("http://") + Server.client().localIP().toString() + String("/"));
  Server.send(302, "text/plain", "");
  Server.client().flush();
  Server.client().stop();
}

void setup() {
  delay(1000);
  Serial.begin(115200);
  Serial.println();
  delay (500);
  File file33 = SPIFFS.open("/dir/tz.txt", "r");
  while (file33.available()){
  tz1 = file33.read();
  }
  Serial.println ("time zone read from flash");
  Serial.println("valore letto"+ tz1);
  tz=tz1;
  file33.close();

  // Enable saved past credential by autoReconnect option,
  // even once it is disconnected.
  
  Config.autoReconnect = false;
  Config.hostName = "Portal";
  // Set access point name
  Config.apid = "Setup";
  // Set menu title
  Config.title = "Cunfiguration";
  
  Portal.config(Config);

  // Load aux. page
  Timezone.load(AUX_TIMEZONE);
  // Retrieve the select element that holds the time zone code and
  // register the zone mnemonic in advance.
  AutoConnectSelect&  tz = Timezone["timezone"].as<AutoConnectSelect>();
  for (uint8_t n = 0; n < sizeof(TZ) / sizeof(Timezone_t); n++) {
    tz.add(String(TZ[n].zone));
  }

  Portal.join({ Timezone });        // Register aux. page

  // Behavior a root path of ESP8266WebServer.
  Server.on("/", rootPage);
  Server.on("/start", startPage);   // Set NTP server trigger handler

  Serial.println("Creating portal and trying to connect...");
  // Establish a connection with an autoReconnect option.
  if (Portal.begin()) {
    Serial.println("WiFi connected: " + WiFi.localIP().toString());
    Serial.println(WiFi.getHostname());
  }

}

void loop() {
  Portal.handleClient();

  if (tz!=tz1){
    // salva file con impostazioni timezone
  File  file22 = SPIFFS.open("/dir/tz.txt", "w");
  if (!file22) Serial.println("file not found, creating new file");
  
  while (file22.available()) file22.print(tz);
  tz1=tz;
  Serial.println ("file printed with tz" );
  Serial.println (tz1 + tz);
  file22.close();

  }
  
}

**Comment by Hieromon** [comment]
It's a Q&A, not an issue, move it to the Discussions and hope for the answer.

**Reply by belu40** [reply]
OK! i wrote in the wrong session, i apologize

**Comment by Hieromon** [comment, marked as answer]
@belu40 First, I answer from the `Simple.ino` improvements that you failed.

AutoConnect's custom web page adopts the ESP Arduino core's WebServer class programming model. According to this model, all procedures that should be written by the user are treated as part of the event handler.

It is a typical example of a server-side model. One thing to keep in mind with server-side programs is that the only opportunity to perform the processing you program in your sketch only exists in response to a request from the client. (Of course, in a real server component, a resident process runs in the background, but that's the service, not a server-side model)
AutoConnect's custom web pages assume the ESP Arduino core's WebServer class programming model and will fail if you do not understand how to make application sketches using the WebServer class. AutoConnect's custom web pages do not replace the server-side programming model with a simple standalone procedural model.

But the documentation of the ESP8266 Arduino core and the ESP32 Arduino core does not explain it clearly, which may cause confusion to the implementors of the sketches. To compensate for that a bit, I described it as a custom web page programming model in the [AutoConnect documentation](https://hieromon.github.io/AutoConnect/achandling.html). Please refer it.

Turning to the your sketch modification, I find that the modification does not conform to the server-side programming model. Your approach is to detect a change in time zone selection and save it to SPIFFS, but **this kind of task should be done in an event handler**. (i.e. it's a custom web page handler)

1. The appropriate procedures where the selected time zone should be saved and restored.
    The function that receives the time zone selected in a custom web page is `startPage()`. Therefore, `startPage` function is responsible for saving the time zone to SPIFFS. On the other hand, the actual setting of the time zone by the `configTime` function should be done when the module is started too. And it is your requirement that the saved time zone should be read out only at the boot time of the ESP module.

2. Make the `configTime` treatment independent.
    It is derived from your request, when the module is started, and when the time zone is changed by the UI. This means that `configTime` execution should be the function, and it is then reasonable to call that function from both `setup()` and `startPage()` functions.

So, we have the modification that you want as followings:

1. Extract the part related to the `configTime` setting from the `startPage()` and make it as a function.

    ```cpp
    void setTime(const String& timezone) {
      for (uint8_t n = 0; n < sizeof(TZ) / sizeof(Timezone_t); n++) {
        String  tzName = String(TZ[n].zone);
        if (timezone.equalsIgnoreCase(tzName)) {
          configTime(TZ[n].tzoff * 3600, 0, TZ[n].ntpServer);
          Serial.println("Time zone: " + timezone);
          Serial.println("ntp server: " + String(TZ[n].ntpServer));
          break;
        }
      }
    }
    ```

2. In `startPage()`, save to SPIFFS and set time zone. To set the time zone, simply call `setTime()` as described above.

    ```cpp
    void startPage() {
      // Retrieve the value of AutoConnectElement with arg function of WebServer class.
      // Values are accessible with the element name.
      String  tz = Server.arg("timezone");
    
      // for (uint8_t n = 0; n < sizeof(TZ) / sizeof(Timezone_t); n++) {
      //   String  tzName = String(TZ[n].zone);
      //   if (tz.equalsIgnoreCase(tzName)) {
      //     configTime(TZ[n].tzoff * 3600, 0, TZ[n].ntpServer);
      //     Serial.println("Time zone: " + tz);
      //     Serial.println("ntp server: " + String(TZ[n].ntpServer));
      //     break;
      //   }
      // }
    
      // Save selected timezone
      File  tzFile = SPIFFS.open("/tz.txt", FILE_WRITE);
      if (tzFile) {
        tzFile.print(tz);
        tzFile.close();
      }
      // Restore timezone configuration
      setTime(tz);
    
      // The /start page just constitutes timezone,
      // it redirects to the root page without the content response.
      Server.sendHeader("Location", String("http://") + Server.client().localIP().toString() + String("/"));
      Server.send(302, "text/plain", "");
      Server.client().flush();
      Server.client().stop();
    }
    ```

3. Restores saved time zone settings with `setup()`.

    ```cpp
    void setup() {
      ...
      // Recover timezone setting
      SPIFFS.begin();
      File  tzFile = SPIFFS.open("/tz.txt", FILE_READ);
      if (tzFile) {
        setTime(tzFile.readString());
        tzFile.close();
      }
    
      // Establish a connection with an autoReconnect option.
      if (Portal.begin()) {
        Serial.println("WiFi connected: " + WiFi.localIP().toString());
      }
    }
    ```

4. Revert `loop()`

    ```cpp
    void loop() {
      Portal.handleClient();
    }
    ```

That's it.

**Reply by Hieromon** [reply]
In addition, You can match the Timezone element of the custom web page with the saved timezone using [AutoConnectSelect::select](https://hieromon.github.io/AutoConnect/apielements.html#select) function like as the followings:

```cpp
void setTime(const String& timezone) {
  for (uint8_t n = 0; n < sizeof(TZ) / sizeof(Timezone_t); n++) {
    String  tzName = String(TZ[n].zone);
    if (timezone.equalsIgnoreCase(tzName)) {
      configTime(TZ[n].tzoff * 3600, 0, TZ[n].ntpServer);
      Serial.println("Time zone: " + timezone);
      Serial.println("ntp server: " + String(TZ[n].ntpServer));

      // Selects the current timezone appropriately
      Timezone["timezone"].as<AutoConnectSelect>().select(timezone);

      break;
    }
  }
}
```

Refer to [AutoConnectSelect API](https://hieromon.github.io/AutoConnect/apielements.html#autoconnectselect) documentation.

**Reply by belu40** [reply]
Thank you very much, i'll try this and let you know!

**Comment by Hieromon** [comment]
@belu40 
>  Then when the network is not available anymore i need to get the access point mode working again after a set timeout but i can't get this working. 

There is already documentation about it.
Refer: https://hieromon.github.io/AutoConnect/adcpcontrol.html#launch-the-captive-portal-on-demand-at-losing-wifi

**Reply by belu40** [reply]
Ok Thanks!

**Comment by belu40** [comment]
Ok, I tried with the suggestions but no success... i think the problem is in the setup() 
now the code is :
 delay(1000);
  Serial.begin(115200);
  Serial2.begin(115200UL);
  SimpleComm.begin();
  Serial.println();
  delay (500);


  // Enable saved past credential by autoReconnect option,
  // even once it is disconnected.
  
  Config.autoReconnect = true;

  //abilita Access point se perde connessione al wifi
  Config.autoRise = true;
  Config.retainPortal = true;

  //altre config  
  Config.hostName = "H228Portal";
  // Set access point name
  Config.apid = "H228_Setup";
  // Set menu title
  Config.title = "Resta";

  //abilita upload programma da rete
  Config.ota = AC_OTA_BUILTIN;
  
  Portal.config(Config);

 // Recover timezone setting
  SPIFFS.begin();
  File  tzFile = SPIFFS.open("/tz.txt", FILE_READ);
  if (tzFile) {
    setTime(tzFile.readString());
    tzFile.close();
  }
  // Load aux. page
  Timezone.load(AUX_TIMEZONE);
  // Retrieve the select element that holds the time zone code and
  // register the zone mnemonic in advance.
  AutoConnectSelect&  tz = Timezone["timezone"].as<AutoConnectSelect>();
  for (uint8_t n = 0; n < sizeof(TZ) / sizeof(Timezone_t); n++) {
    tz.add(String(TZ[n].zone));
  }

  Portal.join({ Timezone });        // Register aux. page

  // Behavior a root path of ESP8266WebServer.
  Server.on("/", rootPage);
  Server.on("/start", startPage);   // Set NTP server trigger handler 

  Serial.println("Creating portal and trying to connect...");
  // Establish a connection with an autoReconnect option.
  if (Portal.begin()) {
    Serial.println("WiFi connected: " + WiFi.localIP().toString());
    Serial.println(WiFi.getHostname());
  }

}

the problem is that i think that the part from //load aux page is not good... because for some reason it doesn't receive the timezone received from spiffs... am i correct?

many thanks in advance

**Reply by belu40** [reply]
i've got this error SPIFFS: mount failed, -10025

**Reply by Hieromon** [reply]
> SPIFFS: mount failed, -10025

It means that SPIFFS is not initialized in advance. Use `SPIFFS.begin(true);`.
Refs: https://github.com/espressif/arduino-esp32/issues/638

**Reply by belu40** [reply]
Ok! Excuse but it seems that my google seaches doesn’t work very well 😓 
Thank you very much! 

Do you have any experience about modbus protocol with your AutoConnect? 
because my entire project is to try to receive some data from an arduino board with serial2 comunication(SimpleComm library ) and then show that data on the root page of autoconnect while sharing them with modbus ( it works with master or slave… i think is better to use slave configuration this time )

Regarding the simpleComm library i have the problem that  if i receive the data packet in the loop and print that in serial monitor i can only see rows of special caracters ( even if i set correctly the baud rate) 

It could be the same as spiffs problem? Have i to write a new class like the One i now use to set time? 

Thank you in advance!

**Reply by Hieromon** [reply]
Do you mean daisy-chaining multiple Arduino boards (e.g. UNO) via RS-485 and connecting the UNO at the end of the chain with Serial2 of ESP32?
If so, I have no experience with the SimleComm library. As for the choice of libraries to use, I think Modbus Master and Modbus Slave have a better track record, don't they?

**Reply by belu40** [reply]
I explain better. I have One of this PLC like modules, this spanish company has done a great job creating a PLC for automation and industrial purposes with many built in features here the link: 
https://www.industrialshields.com/shop/product/007001001300-plc-arduino-ardbox-20-i-os-relay-hf-modbus-wifi-bluetooth-le-1185#attr=815,985,666,849,547,136,346,940,1653,379,65,1370,745,56,1830,1798

I have One of this modules and it’s in specific an integrated arduino leonardo and esp32 built in the module toghether… with optoinsulated input and outputs (also you can use 12 or 24 volts of power supply instead of 5)

i think that your library is very useful to configure the esp inside this module, but i also have to receive signals from the arduino. 
the spanish company sells this modules with a built in serial connection on serial2 and they suggest to use SimpleComm from a packet of
Libraries ( tools40) very useful with this kind of modules. So i hope to finish my project and try to use your library and also exchange the signals, which is the final purpose of my project. 
so the steps are:
1) read and write i/o with built in arduino(i’ve Done that and no problem with this) 
2) create a data packet of variables to send with SimpleComm in serial2 comunication from the arduino to esp32
3) receive data packet with esp32 which needs to be configured with autoConnect to a network
4) share variables of the data packet with modbus tcp once connected to the Wi-Fi network 
5) offer a web page to see real time changing of the variables that are shared with modbus at the same time 

so this is the explanation of the entire project… I hope that it’s clear 
Thanks again!!

**Reply by Hieromon** [reply]
Uh, I have no experience with the industrial Arduino you presented, but I don't think MODBUS has much to do with the cause of the problem. So I came up with some advice to help you diagnose the problem you encountered. Here they are:

1. Is communication successful between ESP32 Serial2 and Leonardo using a simple sketch without AutoConnect? The garbled characters are mostly caused by the RS232 bit rate not matching or the LSB and MSB being reversed. Even if the baud rate is the same, garbled characters will occur if the stop bit and parity bit of the communication partner is mismatched.

2. If the receiving from Serial2 is done within the loop function, then the code would normally look like the following:

```cpp
void loop() {
  while (Serial2.available()) {
    char data = Serial2.read();
    Serial.write(data);
  }
}
```

However, if the following additional modification is made to the above code to parallelize AutoConnect::handleClient(), it will not work properly. This is a common mistake.

```cpp
void loop() {

  portal.handleClient();

  while (Serial2.available()) {
    char data = Serial2.read();
    Serial.write(data);
  }
}
```

AutoConnect is an event handler for the Web server class, so in an extreme case, it is a mass of ISR for TCP interrupts. So, the above changes cannot accommodate actual hardware events in the following two ways:

- The Serial2 receive buffer may overflow between the time the web server class receives the TCP packet and the time AutoConnect completes processing.

- On the other hand, wrapping the processing of Serial2 received data in a while loop will interfere with the call to AutoConnect::handleClient, which will delay or time out the TCP response.

I don't have any experience with the [SimpleComm](https://github.com/IndustrialShields/arduino-SimpleComm) library you mentioned, but I found it after exploring the GitHub repository. So the [SimleComClass::receive](https://github.com/IndustrialShields/arduino-SimpleComm/blob/5a322d7b5e4db1c0364f089b3fd3724a9caceb5b/src/SimpleComm.cpp#L98) function forms a while loop for receiving, and it clearly falls under the above mistake.

To perform TCP receive handling and Serial2 receive handling in parallel, use the multi-core that is a feature of ESP32, and do one of them in another task.

```cpp
#include <Arduino.h>
#include <WiFi.h>
#include <WebServer.h>
#include <AutoConnect.H>

// Serial2 configuration, attention to data bit width, stop bit width, and parity.
const unsigned long _serial2Baudrate = 9800;
const uint32_t  _uart2Config = SERIAL_8N1;
const int8_t    _rx2Pin =  16;
const int8_t    _tx2Pin =  17;

// AutoConnect task configuration
TaskHandle_t  taskAutoConnect;
const BaseType_t  taskRunningCore = (1 - CONFIG_ARDUINO_RUNNING_CORE);
void  vTaskAutoConnect(void* pvParameters);

AutoConnect portal;
AutoConnectConfig config;

void setup() {
  delay(1000);
  Serial.begin(115200);
  Serial.println();

  Serial2.begin(_serial2Baudrate, _uart2Config, _rx2Pin, _tx2Pin);

  config.autoReconnect = true;
  config.reconnectInterval = 1;
  config.retainPortal = true;
  config.ota = AC_OTA_BUILTIN;
  config.tickerOn = HIGH;
  config.tickerPort = 2;
  portal.config(config);
  portal.begin();

  xTaskCreatePinnedToCore(vTaskAutoConnect, "vTaskAutoConnect", 1024 * 8, NULL, 0, &taskAutoConnect, taskRunningCore);
}

void vTaskAutoConnect(void* pvParameters) {
  while (true)
    portal.handleClient();
}

void loop() {
  while (Serial2.available()) {
    char data = Serial2.read();
    Serial.write(data);
  }
  delay(1);
}
```

**Reply by belu40** [reply]
Ok! thank you again! i will try that!
now my loop code looks like that:
void loop() {
  Portal.handleClient();


    if (SimpleComm.receive(Serial2, packet)) {
    const data_t *data = (const data_t *) packet.getData();
    if (data != nullptr) {
      // Print received data
      Serial.print("Received data: ");
      Serial.print(data->name);
      Serial.print("; ");
      timeDone = data->time;
      Serial.print(timeDone);
      Serial.print("; ");
      doneCounter = data->counter;
      Serial.print(doneCounter);
      Serial.print("; ");
      workState = data->state;
      Serial.print(workState);
      Serial.println();
    }
  }
}

so it's clear that the error i made is to parallelize my code to the portal.handleClient(). I will try to work with multicore functions and let you know!.

i receive the data from serial2 once every 2/3 seconds because i send them from leonardo every 2/3 seconds. so i think i can try to use one core to either receive from serial2 either send data in modbus when are received. then i use another core to handle the client
I'll let you know!

**Reply by belu40** [reply]
in my case i've already tried to use this code without autoconnect and it works, so i think the problem can be really solved by using another core!

**Reply by belu40** [reply]
Hello!
i've tried this one!
It's all good until i try to connect to the portal root page after the first configuration in ap mode.
I'm able to connect to AP and configure the network. when it receives the ip address i try to connect to it but the page is shown only once, ad if i try to reload it does't work anymore.

Do you think is possible that only 1 core for AutoConnect is not enough powerful to run the portal? or am i wrong with something else?

in any case thanks again!

I can see forward for the finish flag for this project!

**Reply by Hieromon** [reply]
Try this one. Made the running core of AutoConnect::handleClient the same as the core of the loop function, and moved the Serial2 running core to core 0.

```cpp
#include <Arduino.h>
#include <WiFi.h>
#include <WebServer.h>
#include <AutoConnect.H>

// Serial2 monitoring page
const char  PAGE_SERIAL2_MONITOR[] = R"*(
{
  "title": "Serial2",
  "uri": "/",
  "menu": true,
  "element": [
    {
      "name": "serial2data",
      "type": "ACText",
      "format": "<div style='font-weight:bold'>%s</div>"
    },
    {
      "name": "reload",
      "type": "ACSubmit",
      "value": "Reload",
      "uri": "/"
    }
  ]
}
)*";


// Serial2 configuration, attention to data bit width, stop bit width, and parity.
const unsigned long _serial2Baudrate = 9800;
const uint32_t  _uart2Config = SERIAL_8N1;
const int8_t    _rx2Pin =  16;
const int8_t    _tx2Pin =  17;

// AutoConnect task configuration
TaskHandle_t  taskSerial2;
const uint32_t    taskStackDepth = (8 * 1024);
const BaseType_t  taskRunningCore = (1 - CONFIG_ARDUINO_RUNNING_CORE);
void  vTaskSerial2(void* pvParameters);

AutoConnect portal;
AutoConnectConfig config;
AutoConnectAux  serial2;

void setup() {
  delay(1000);
  Serial.begin(115200);
  Serial.println();

  Serial2.begin(_serial2Baudrate, _uart2Config, _rx2Pin, _tx2Pin);

  config.autoReconnect = true;
  config.reconnectInterval = 1;
  config.retainPortal = true;
  config.ota = AC_OTA_BUILTIN;
  config.tickerOn = HIGH;
  config.tickerPort = 2;
  portal.config(config);

  serial2.load(PAGE_SERIAL2_MONITOR);
  portal.join({ serial2 });

  portal.begin();

  xTaskCreatePinnedToCore(vTaskSerial2, "vTaskSerial2", taskStackDepth, NULL, 0, &taskSerial2, taskRunningCore);
}

void vTaskSerial2(void* pvParameters) {
  while (true) {
    while (Serial2.available()) {
      char  data = Serial2.read();
      Serial.write(data);
      String  serial2Data = serial2["serial2data"].as<AutoConnectText>().value + String(data);
      if (serial2Data.length() > 32)
        serial2Data = serial2Data.substring(1, 33);
      serial2["serial2data"].as<AutoConnectText>().value = serial2Data;
    }

    // Allow CPU to switch to other tasks.
    delay(1);
  }
}

void loop() {
  portal.handleClient();

  // Allow CPU to switch to other tasks.
  delay(1);
}
```

**Reply by belu40** [reply]
Hello!
I made great progresses with this example.
i got this error on serial monitor when  i reboot the ESP
```
Rebooting...
ets Jun  8 2016 00:22:57

rst:0xc (SW_CPU_RESET),boot:0x13 (SPI_FAST_FLASH_BOOT)
configsip: 0, SPIWP:0xee
clk_drv:0x00,q_drv:0x00,d_drv:0x00,cs0_drv:0x00,hd_drv:0x00,wp_drv:0x00
mode:DIO, clock div:1
load:0x3fff0018,len:4
load:0x3fff001c,len:1216
ho 0 tail 12 room 4
load:0x40078000,len:10944
load:0x40080400,len:6388
entry 0x400806b4

Time zone: Europe/Rome
ntp server: europe.pool.ntp.org
Guru Meditation Error: Core  1 panic'ed (LoadProhibited). Exception was unhandled.
Core 1 register dump:
PC      : 0x400e0906  PS      : 0x00060530  A0      : 0x800d196c  A1      : 0x3ffb1e70  
A2      : 0x00000008  A3      : 0x3ffb1f20  A4      : 0x00000000  A5      : 0x00000000  
A6      : 0x00000000  A7      : 0x3ffc6ac4  A8      : 0x800ef8a6  A9      : 0x3ffb1e50  
A10     : 0x3ffb1ec0  A11     : 0x3f400903  A12     : 0x00000008  A13     : 0x0000ff00  
A14     : 0x00ff0000  A15     : 0xff000000  SAR     : 0x00000008  EXCCAUSE: 0x0000001c  
EXCVADDR: 0x00000020  LBEG    : 0x400014fd  LEND    : 0x4000150d  LCOUNT  : 0xfffffffd  

ELF file SHA256: 0000000000000000

Backtrace: 0x400e0906:0x3ffb1e70 0x400d1969:0x3ffb1e90 0x400d21f6:0x3ffb1ef0 0x400f0696:0x3ffb1fb0 0x40089b06:0x3ffb1fd0
```
it could be caused by saving timezone to spiffs?
the error doesn't occurres when i din't set any timezone before shutting down the power supply.

This is part of the code 
```c_cpp
void setTime(const String& timezone) {
  for (uint8_t n = 0; n < sizeof(TZ) / sizeof(Timezone_t); n++) {
    String  tzName = String(TZ[n].zone);
    if (timezone.equalsIgnoreCase(tzName)) {
      configTime(TZ[n].tzoff * 3600, 0, TZ[n].ntpServer);
      Serial.println("Time zone: " + timezone);
      Serial.println("ntp server: " + String(TZ[n].ntpServer));

      // Selects the current timezone appropriately
      Timezone["timezone"].as<AutoConnectSelect>().select(timezone);

      break;
    }
  }
}

void startPage() {
  // Retrieve the value of AutoConnectElement with arg function of WebServer class.
  // Values are accessible with the element name.
 String tz = Server.arg("timezone");

  // Save selected timezone

  File  tzFile = SPIFFS.open("/tz.txt", FILE_WRITE);
  if (tzFile) {
    tzFile.print(tz);
    tzFile.close();
     Serial.println("TZ saved"+tz);
  }
 
  // Restore timezone configuration
  setTime(tz);

  // The /start page just constitutes timezone,
  // it redirects to the root page without the content response.
  Server.sendHeader("Location", String("http://") + Server.client().localIP().toString() + String("/"));
  Server.send(302, "text/plain", "");
  Server.client().flush();
  Server.client().stop();
}
```

and this is in the void setup() function:
```clike
  SPIFFS.begin(true);
  File  tzFile = SPIFFS.open("/tz.txt", FILE_READ);
  if (tzFile) {
    setTime(tzFile.readString());
    tzFile.close();
  }
  // Load aux. page
  Timezone.load(AUX_TIMEZONE);
  // Retrieve the select element that holds the time zone code and
  // register the zone mnemonic in advance.
  AutoConnectSelect&  tz = Timezone["timezone"].as<AutoConnectSelect>();
  for (uint8_t n = 0; n < sizeof(TZ) / sizeof(Timezone_t); n++) {
    tz.add(String(TZ[n].zone));
  }

  Portal.join({ Timezone });        // Register aux. page

  // Behavior a root path of ESP8266WebServer.
  Server.on("/", rootPage);
  Server.on("/start", startPage);   // Set NTP server trigger handler 

  Serial.println("Creating portal and trying to connect...");
  // Establish a connection with an autoReconnect option.
  if (Portal.begin()) {
    Serial.println("WiFi connected: " + WiFi.localIP().toString());
    Serial.println(WiFi.getHostname());
  }
```
Hope to find from you the solution and suggestions also for this problem!
thank you a lot!

**Reply by Hieromon** [reply]
@belu40 The problem cause lies in the process sequence of the `setup` with your modification. You may have gotten a little lost, since the [comment](https://github.com/Hieromon/AutoConnect/discussions/454#discussioncomment-2059935) representing my advice does not clearly indicate when to call the `setTime` from the `setup`.

The call to the `setTime` function from the `setup` function must be made after the *Timezone* AutoConnectAux is loaded. In the above [comment](https://github.com/Hieromon/AutoConnect/discussions/454#discussioncomment-2059935), the start of SPIFFS and the call to setTime are made after *Timezone* AutoConnectAux is loaded and just before `portal.begin()`, which is the appropriate timing.

On the other hand, according to your modification, the `setTime` call is made before the *Timezone* AutoConnectAux is loaded. That is, the `setTime` function will refer `Timezone["timezone"].as<AutoConnectSelect>()` at a time when **no instance** of the `"timezone"` element exists. So this is the cause of the `LoadProhibited` exception.
